KIPs/kip-149.md

```diff
@@ -0,0 +1,274 @@
+
+#### Chain Configuration
+
+In the Chain Config, the following field is introduced. All node operators in a network must update `genesis.json` configuration with the same value. The configuration values for Baobab and Cypress networks are hard-coded on the client source code.
+
+- `kip149CompatibleBlock`: the target block number that a Klaytn node deploys Registry.
+
+#### Execution
+
+The Registry deployment is executed at the `Finalize` function, which means the end of the block processing process. It reads the reserved address and runtime bytecode and deploys the Registry by the bytecode injection.
+
+```go
+// Deploy the registry contract if the block is the kip-149 fork block
+if chain.Config().IsKIP149ForkBlock(header.Number) {
+	// Set the registry contract code to the stateDB
+	err := registry.InstallRegistry(state)
+	if err != nil {
+		logger.Error("failed to set the registry contract code", "err", err)
+	} else {
+		logger.Info("successfully set the registry contract code", "block", header.Number.Uint64())
+	}
+}
+```
+
+#### Resolver
+
+A Klaytn node will have a resolver to read the active addresses of system contracts from the Registry.
+
+## Rationale
+
+### Bytecode Injection For Registry Deployment
+
+In the future, all system contracts will be registered in the Registry, and a Klaytn node will read the active addresses of system contracts from the Registry. Not only for a Klaytn node but also for other ecosystem participants who will use the registry to read the system contracts they need, meaning the registry should be registered at an easily accessible reserved address.
+
+### Separate Data And Logic Contract
+
+This proxy pattern simplifies the process of system contract update because the existing data can be used even if the logic contract is changed. The main issue of the proxy pattern is the centralization and potential private key loss. But delegating ownership to upgrade a logic contract to Governance can solve those problems.
+
+## Backward Compatibility
+
+### Deployed System Contracts
+
+As explained in [Registry](#1-registry) section, it consists of base and network-specific registries. The below code is an example of `RegistryCypress`.
+
+```solidity
+// Pre-deployed system contracts
+address public constant ADDRESS_BOOK = 0x0000000000000000000000000000000000000400;
+address public constant VOTING = 0xcA4Ef926634A530f12e55A0aEE87F195A7B22Aa3;
+address public constant TREASURY_REBALANCE = 0xD5ad6D61Dd87EdabE2332607C328f5cc96aeCB95;
+address public constant STAKING_TRACKER = 0x9b8688d616D3D5180d29520c6a0E28582E82BF4d;
+
+bytes32 public constant ADDRESS_BOOK_KEY = keccak256(abi.encodePacked("AddressBook"));
+bytes32 public constant VOTING_KEY = keccak256(abi.encodePacked("Voting"));
+bytes32 public constant TREASURY_REBALANCE_KEY = keccak256(abi.encodePacked("TreasuryRebalance"));
+bytes32 public constant STAKING_TRACKER_KEY = keccak256(abi.encodePacked("StakingTracker"));
```

Comment:
Is the `GovParam` contract intentionally excluded from management within the `Registry` contract?

Reply:
I'm curious about whether `GovParam` is deployed or not. If deployed, then we need to include it in the wrapper. But if not, we can register `GovParam` in the `register` function when needed.

Reply:
This comment is not valid at this point as the usage was revised.

The `GovParam` contract is currently deployed on both Cypress and Baobab netowrk. You can query the address through `klay.getParams()`.